### PR TITLE
ENH: Extending CPU feature detection framework to support IBM Z SIMD

### PR DESCRIPTION
**Pull request by pradghos**
We would like to extend CPU feature detection infrastructure for IBM Z CPU features. 
Eventually It will help to add and enable Z specific SIMD builtins and instructions.

As part of the PR, we have extended `ccompiler_opt.py`, `npy_cpu_features.c.src`, 
`npy_cpu_features.h`, other files for Z CPU feature detection and added test files for 
`VX/VXE/VXE2` in `distutils/checks`.

cc @edelsohn @Andreas-Krebbel @potula-chandra @ajaypvictor

### CODE REVIEW

**Comment by edelsohn** [comment]
@seiko2plus Sayed, what's your progress?

File: numpy/core/src/common/npy_cpu_features.c.src [hunk 1]
```diff
@@ -509,6 +510,42 @@ npy__cpu_init_features(void)
 #endif
 }
 
+/***************** ZARCH ******************/
+
+#elif defined(NPY_CPU_S390) || defined(__s390x__)
```
Comment (seiko2plus) [inline numpy/core/src/common/npy_cpu_features.c.src:515]:
```suggestion
#elif defined(__s390x__)
```
it supposed be enabled only for zarch

**Review by seiko2plus — CHANGES_REQUESTED** [review]
Hello @pradghos, along with the following suggestions we still need to add to test runtime detection similar to:
https://github.com/numpy/numpy/blob/93301a53aad8ddda15a25415e9af8f574cd758a8/numpy/core/tests/test_cpu_features.py#L140-L147.

**Comment by seiko2plus** [comment]
@edelsohn,
> @seiko2plus Sayed, what's your progress?

we have several SIMD kernels now written by universal intrinsics, and more coming. I was working on s390x implementation but I pause it in favor of replacing/implementing more kernels to universal intrinsics first.

**Comment by pradghos** [comment]
Hi @seiko2plus,  I have incorporated the review comments now, Thank you very much for the review !

**Review by seiko2plus — APPROVED** [review]
Thank you Pradipta, I have updated the document of build options, and added extra tests for s390x to guarantee that IBM Z features don't get affected by future changes to our infrastructure.

**Comment by mattip** [comment]
Thanks @pradghos, @seiko2plus

**Comment by potula-chandra** [comment]
Excellent thank you @pradghos, @seiko2plus  and  @mattip

File: numpy/distutils/tests/test_ccompiler_opt.py [hunk 2]
```diff
@@ -432,14 +435,15 @@ def test_args_options(self):
...
+            "sse sse2 vsx vsx2 neon neon_fp16 vx vxe",
             x86_gcc="-msse -msse2", x86_icc="-msse -msse2",
             x86_iccw="/arch:SSE2",
             x86_msvc="/arch:SSE2" if self.march() == "x86" else "",
             ppc64_gcc= "-mcpu=power8",
             ppc64_clang="-maltivec -mvsx -mpower8-vector",
             armhf_gcc="-mfpu=neon-fp16 -mfp16-format=ieee",
-            aarch64=""
+            aarch64="",
+            s390="-mzvector -march=arch12"
```
Comment (seiko2plus) [inline numpy/distutils/tests/test_ccompiler_opt.py:446]: what a typo see #20588

**Comment by InessaPawson** [comment]
Hi-five on merging your first pull request to NumPy, @pradghos! We hope you stick around! Your choices aren’t limited to programming – you can review pull requests, help us stay on top of new and old  issues, develop educational material, work on our website, add or improve graphic design, create marketing materials, translate website content, write grant proposals, and help with other fundraising initiatives. For more info, check out: https://numpy.org/contribute/. 
Also, consider joining our mailing list. This is a great way to connect with other cool people in our community and be part of important conversations that affect the development of NumPy: https://mail.python.org/mailman/listinfo/numpy-discussion

**Pull request merged**